### ISSUE TITLE
Emit variable is set to `any`, breaks build but works fine in dev mode

### ISSUE THREAD

**Issue by rodrigocfd** [vuejs/language-tools#3252]
### Vue version

3.3.4

### Link to minimal reproduction

https://github.com/rodrigocfd/vue-bug-emits

### Steps to reproduce

Run `npm run dev` and everything works fine.

Run `npm run build` and it won't compile.

### What is expected?

A successful build with `npm run build`.

### What is actually happening?

With `npm run build`, the following error shows up:

```
src/App.vue:10:22 - error TS7006: Parameter 'n' implicitly has an 'any' type.

10  <Child @changeName="n => name = n" />
                        ~


Found 1 error in src/App.vue:10
```

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
    Memory: 25.86 GB / 31.95 GB
  Binaries:
    Node: 18.14.1 - C:\Program Files\Nodejs\node.EXE
    npm: 9.6.7 - C:\Program Files\Nodejs\npm.CMD
  Browsers:
    Chrome: 114.0.5735.91
    Edge: Spartan (44.19041.1266.0)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    vue: ^3.3.4 => 3.3.4
```

### Any additional comments?

I tested the original dependencies installed with `npm create vite@latest`, and then updating all dependencies. The behavior is the same.

Curiously, despite running flawlessly in dev mode, when hovering the variable in VSCode, Volar reports the emit parameter as `any`, which is exactly the build error:

![volar-any](https://github.com/vuejs/core/assets/1283868/92a23918-253b-4674-ab42-9d4d13f6c1c7)

**Comment by pietdevries94** [comment]
Have you recently updated your codebase to TS 5.1? Because that's where I started to have these problems too. As a work-around for now you can set your typescript to "~5.0.4".

Here is a fork of the reproduction repository, with TS on 5.0.4 as only change. The build command works perfectly here. Volar also picks up the emit types correctly.

TL;DR: TS 5.1 seems to break the correct types of emits

**Comment by pietdevries94** [comment]
dev mode doesn't use the language tools. It just strips TS away and runs. The build actually checks your TS.

If you set your workspace to use the TS from your node_modules and you have reloaded your vue language server (or your complete vs code window) after installin 5.0, it should also correctly show the typings again in vs code

**Comment by LinusBorg** [comment]
> @LinusBorg How is this an "implicit any", if I'm explicitly typing my emit in `Child`?

You're right, the correct way to look at this is that Vue Language Tools / vue-tsc should infer the argument from the emit type.

**Comment by so1ve** [comment]
Related: #3268
